scheme/src/main/java/cz/o2/proxima/scheme/confluent/SchemaRegistryValueSerializer.java

```diff
@@ -0,0 +1,207 @@
+/**
+ * Copyright 2017-2019 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.scheme.confluent;
+
+import com.google.common.annotations.VisibleForTesting;
+import cz.o2.proxima.scheme.SerializationException;
+import cz.o2.proxima.scheme.ValueSerializer;
+import cz.o2.proxima.scheme.avro.AvroSerializer;
+import cz.o2.proxima.util.Classpath;
+import io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient;
+import io.confluent.kafka.schemaregistry.client.SchemaMetadata;
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericContainer;
+import org.apache.avro.specific.SpecificRecord;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Slf4j
+class SchemaRegistryValueSerializer<M extends GenericContainer>
+    implements ValueSerializer<M> {
+
+  static final byte MAGIC_BYTE = 0x0;
+  static final int SCHEMA_ID_SIZE = 4;
+  private final URI schemaRegistryUri;
+  private transient SchemaRegistryClient schemaRegistry = null;
+  private Class<M> clazz = null;
+  private Integer schemaId = null;
+
+  private transient Map<Integer, AvroSerializer> serializersCache = null;
+
+
+  SchemaRegistryValueSerializer(URI scheme) throws URISyntaxException {
+    this.schemaRegistryUri = new URI(scheme.getSchemeSpecificPart());
+  }
+
+  @Override
+  public Optional<M> deserialize(byte[] input) {
+    if (input.length == 0) {
+      /**
+       * This is a little bit weird but proxima try to deserialize empty bytes
+       * in serializer validation {@link ValueSerializer#isValid(byte[])}
+       */
+      return Optional.of(getDefault());
+    }
+
+    try {
+      return deserializeValue(input);
+    } catch (Exception e) {
+      throw new SerializationException("Unable to deserialize value.", e);
+    }
+  }
+
+  @Override
+  public byte[] serialize(M value) {
+    try {
+      return serializeValue(value, getSchemaId());
+    } catch (IOException e) {
+      throw new SerializationException("Unable to serialize data.", e);
+    }
+  }
+
+  @Override
+  public M getDefault() {
+    try {
+      return getClassType().newInstance();
```

Comment:
It might be better idea to cache this, because it might be called often.

Reply:
Fixed